File: Striver/DP/Q35.java
package Striver.DP;

public class Q35 {
    // best time to buy and sell stock 

    public static int maxProfit(int[] prices) {
        int n = prices.length;
        int[] dp = new int[n];
        dp[0] = 0;
        int min = prices[0];
        for(int i=1; i<n; i++){
            min = Math.min(min, prices[i]);
            dp[i] = Math.max(dp[i-1], prices[i]-min);
        }
        return dp[n-1];
    }

    public static void main(String[] args) {
        int[] prices = {7,1,5,3,6,4};
        System.out.println(maxProfit(prices));
    }
}

File: Striver/DP/Q35.py
# best time to buy and sell stock

def maxProfit(prices):
    n = len(prices)
    dp = [0] * n
    minPrice = prices[0]
    for i in range(1, n):
        dp[i] = max(dp[i-1], prices[i]-minPrice)
        minPrice = min(minPrice, prices[i])

    return dp[n-1]

prices = [7,1,5,3,6,4]
print(maxProfit(prices))